### PR TITLE
Express Notes API

### PR DESCRIPTION
**Pull request by gavenlou**
### GET
image
image
image
image
### POST
image
image
### DELETE
image
image
image
### PUT
image
image
image
image
image

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+app.use(parse);
+
+app.get('/api/notes', (req, res) => {
+  res.status(200);
+  let reply = [];
+  for (let key in data.notes) {
+    reply.push(data.notes[key]);
+  }
+  return res.send(reply);
+})
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: "Clients can GET a list of notes" looks good.

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+app.get('/api/notes/:id/', (req, res) => {
+  let reply = [];
+  for (let key in data.notes) {reply.push(key);}
+  if (req.params.id < 0 || isNaN(req.params.id)) {
+    res.status(400)
+    return res.json({"error": "id must be a positive integer."});
+  }
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: `400 Bad Request` should also trigger on fractional `id`s.

---

image

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  data.notes[data.nextId].id = data.nextId;
+  data.nextId++;
+  fs.writeFile('data.json', JSON.stringify(data, null, 2) , (err) => {
+    if (err) {
+      res.status(500);
+      return res.json({ "error": "An unexpected error occurred."});
+    }
+    else return res.send(req.body);
+  })
+})
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: "Clients can POST a new note" looks good.

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+app.delete('/api/notes/:id', (req, res) => {
+  let reply = [];
+  for (let key in data.notes) { reply.push(key); }
+  if (req.params.id < 0 || isNaN(req.params.id)) {
+    res.status(400);
+    return res.json({ "error": "id must be a positive integer."});
+  }
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: Like above, `400 Bad Request` should trigger any time `id` isn't a positive integer.

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  })
+})
+
+app.put('/api/notes/:id', (req, res) => {
+  let reply = [];
+  for (let key in data.notes) { reply.push(key); }
+  if (req.params.id < 0 || isNaN(req.params.id)) {
+    res.status(400);
+    return res.json({ "error": "id must be a positive integer." })
+  }
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: Like above, `400 Bad Request` should trigger for _all_ `id`s which aren't positive integers.

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

**Comment by gavenlou** [comment]
Changed testing for number to testing for Integer.

**Review by TimelikeClosure — APPROVED** [review]

**Pull request merged**